### DISCUSSION TITLE
Keycloak Docker container can't make make connection with Postgres database container: "Failed to obtain JDBC connection", "Connection to postgres:5432 refused".

### DISCUSSION THREAD

**Discussion by nkamp** [keycloak/keycloak#34722] [Q&A]
I'm working with Windows 11.

I have an Docker container for Keycloak and an Postgres database container. I'm working with Docker-compose.yml and Docker-keycloak file
Keycloak has a storage provider which does a request to the Postgres application database.

I can update\save data with the keycloak in the keycloak database like clients etc. So keycloak with keycloak database itself is working.
But If click on users in the keycloak admin console then a request is done to the application database but than I get the error:

    ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-4) The connection attempt failed.

I have in my Dockerfile:
Keycloak: KC_DB_URL=jdbc:postgresql://host.docker.internal:5432/keycloak
and webapp: ____DB_URL=jdbc:postgresql://host.docker.internal:5432/webapp

I have tried ‘localhost:5432’, ‘127.0.0.1:5432’, 'postgres_db:5432' (postgres_db is de service name of the postgres database) but still the same error message.

### Keycloak from the Keycloak container tried to make connection with the webapp database in the Postgres database container. This request fails:"Connection to postgres:5432 refused" .

**Or is this not possible and do I need for this an nginx server?**

Here is my new Docker compose.yml and Dockerfile:
```
services:
  postgres_db:
    image: postgis/postgis:14-master
    ports:
      - "5432:5432"                           # -p
    container_name: yyyy                     # --name
    build:
      context: .
      dockerfile: Dockerfile-db
    restart: always
    environment:
      POSTGRES_DB: xxxxx                      
      POSTGRES_USER: xxxx                      
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD} 
    env_file:
      - .env    
    networks:
      - pg-network
    volumes:
      - ./aaaa_container_db-data/:/var/lib/postgresql/data/          # -v
      - ./aaaa_container_tmp_data:/tmp
      - ./init-multi-postgres-databases.sh:/docker-entrypoint-initdb.d/init-multi-postgres-databases.sh
    healthcheck:
      test: [ "CMD-SHELL", "pg_isready -h localhost -U $$POSTGRES_USER" ]
      interval: 15s
      timeout: 10s
      retries: 12
      
      # here the PgAdmin
      #...
      #...

  auth:
    image: quay.io/keycloak/keycloak:25.0.6
    container_name: keycloak
    restart: unless-stopped
    ports:
      - "8180:8180"
    env_file:
      - .env    # The name of your environment file (the one at the repository root)
    build:
      context: .
      dockerfile: Dockerfile-keycloak
    depends_on:
      postgres_db:
          condition: service_healthy
    networks:
      - pg-network
      
networks:
  pg-network:
    driver: bridge

volumes:
  aaaa_container_db-data:
```

Here is my Dockerfile-keycloak:
````
FROM quay.io/keycloak/keycloak:25.0.6

ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true
ENV KC_DB=postgres

ENV KC_DB_URL=jdbc:postgresql://host.docker.internal:5432/keycloak
ENV KC_HOSTNAME=localhost
ENV KC_DB_USERNAME=xxxx
ENV KC_DB_PASSWORD=yyyy
ENV DEBUG_PORT="8787"
ENV ____DB_URL=jdbc:postgresql://host.docker.internal:5432/appdb
ENV ____DB_USER=xxxx
ENV ____DB_PASSWORD=yyyy
ENV KEYCLOAK_ADMIN=admin
ENV KEYCLOAK_ADMIN_PASSWORD=admin
ENV KEYCLOAK_LOGLEVEL=DEBUG
ENV KC_HOSTNAME_STRICT=false

COPY ./conf/keycloak/gir-keycloak-user-storage.jar /opt/keycloak/providers/gir-keycloak-user-storage.jar

COPY ./conf/keycloak/quarkus.properties /opt/keycloak/conf/quarkus.properties
COPY ./conf/keycloak/brzo-local-realm.json /opt/keycloak/data/import/brzo-local-realm.json

WORKDIR /opt/keycloak

RUN mkdir /opt/keycloak/ObjectStore && \
    chown keycloak:keycloak /opt/keycloak/ObjectStore && \
    chmod 774 /opt/keycloak/ObjectStore

# Install the PostgreSQL JDBC driver
RUN /opt/keycloak/bin/kc.sh build --db=postgres

ENTRYPOINT ["./bin/kc.sh"]
CMD ["--verbose", \
    "start", \
    "--import-realm",  \    
    "--spi-connections-jpa-default-migration-strategy=update", \
    "--spi-events-listener-jboss-logging-success-level=info", \
    "--spi-events-listener-jboss-logging-error-level=warn", \
    "--spi-login-protocol-openid-connect-legacy-logout-redirect-uri=true", \
    "--http-enabled=true", \
    "--http-port=8180", \
    "--proxy=passthrough", \
    "--optimized"]
````

and here is Docker container logging:

> 2024-11-06 21:44:14,177 WARN  [org.keycloak.quarkus.runtime.cli.Picocli] (main) The following used options or option values are DEPRECATED and will be removed or their behaviour changed in a future release:
>         - proxy: Use proxy-headers.
> Consult the Release Notes for details.
> 2024-11-06 21:44:14,652 WARN  [org.keycloak.protocol.oidc.OIDCLoginProtocolFactory] (main) Deprecated switch 'legacy-logout-redirect-uri' is enabled. Please try to disable it and update your clients to use OpenID Connect compliant way for RP-initiated logout.
> 2024-11-06 21:44:16,938 INFO  [io.agroal.pool] (JPA Startup Thread: keycloak-default) Datasource '<default>': Connection reap not supported in pool-less mode
> 2024-11-06 21:44:16,941 INFO  [io.agroal.pool] (JPA Startup Thread: keycloak-default) Datasource 'users': Connection reap not supported in pool-less mode
> 2024-11-06 21:44:17,192 INFO  [org.infinispan.CONTAINER] (ForkJoinPool.commonPool-worker-1) ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'
> 2024-11-06 21:44:17,416 WARN  [org.jgroups.stack.Configurator] (ForkJoinPool.commonPool-worker-1) JGRP000014: ThreadPool.thread_dumps_threshold has been deprecated: ignored
> 2024-11-06 21:44:17,430 INFO  [org.infinispan.CLUSTER] (ForkJoinPool.commonPool-worker-1) ISPN000078: Starting JGroups channel `ISPN` with stack `udp`
> 2024-11-06 21:44:17,434 INFO  [org.jgroups.JChannel] (ForkJoinPool.commonPool-worker-1) local_addr: 11e76b0e-66d6-4b5b-8a24-23ffd7e29b4c, name: e02503325114-10438
> 2024-11-06 21:44:17,439 WARN  [org.jgroups.protocols.UDP] (ForkJoinPool.commonPool-worker-1) JGRP000015: the send buffer of socket MulticastSocket was set to 1MB, but the OS only allocated 212.99KB
> 2024-11-06 21:44:17,441 WARN  [org.jgroups.protocols.UDP] (ForkJoinPool.commonPool-worker-1) JGRP000015: the receive buffer of socket MulticastSocket was set to 20MB, but the OS only allocated 212.99KB
> 2024-11-06 21:44:17,442 WARN  [org.jgroups.protocols.UDP] (ForkJoinPool.commonPool-worker-1) JGRP000015: the send buffer of socket MulticastSocket was set to 1MB, but the OS only allocated 212.99KB
> 2024-11-06 21:44:17,442 WARN  [org.jgroups.protocols.UDP] (ForkJoinPool.commonPool-worker-1) JGRP000015: the receive buffer of socket MulticastSocket was set to 25MB, but the OS only allocated 212.99KB
> 2024-11-06 21:44:17,452 INFO  [org.jgroups.protocols.FD_SOCK2] (ForkJoinPool.commonPool-worker-1) server listening on *.31194
> 2024-11-06 21:44:19,458 INFO  [org.jgroups.protocols.pbcast.GMS] (ForkJoinPool.commonPool-worker-1) e02503325114-10438: no members discovered after 2003 ms: creating cluster as coordinator
> 2024-11-06 21:44:19,465 INFO  [org.infinispan.CLUSTER] (ForkJoinPool.commonPool-worker-1) ISPN000094: Received new cluster view for channel ISPN: [e02503325114-10438|0] (1) [e02503325114-10438]
> 2024-11-06 21:44:19,503 INFO  [org.infinispan.CLUSTER] (ForkJoinPool.commonPool-worker-1) ISPN000079: Channel `ISPN` local address is `e02503325114-10438`, physical addresses are `[172.18.0.4:46730]`
> 2024-11-06 21:44:26,979 WARN  [io.agroal.pool] (JPA Startup Thread: users) Datasource 'users': The connection attempt failed.
> 2024-11-06 21:44:26,979 WARN  [org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator] (JPA Startup Thread: users) HHH000342: **_Could not obtain connection to query metadata: org.hibernate.exception.JDBCConnectionException: unable to obtain isolated JDBC connection_**
>         at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:100)
>         at org.hibernate.resource.transaction.backend.jta.internal.JtaIsolationDelegate.lambda$new$1(JtaIsolationDelegate.java:74)
>         at org.hibernate.resource.transaction.backend.jta.internal.JtaIsolationDelegate.doTheWork(JtaIsolationDelegate.java:202)
>         at org.hibernate.resource.transaction.backend.jta.internal.JtaIsolationDelegate.lambda$delegateWork$3(JtaIsolationDelegate.java:91)
>         at org.hibernate.resource.transaction.backend.jta.internal.JtaIsolationDelegate.doInSuspendedTransaction(JtaIsolationDelegate.java:123)
>         at org.hibernate.resource.transaction.backend.jta.internal.JtaIsolationDelegate.delegateWork(JtaIsolationDelegate.java:88)
>         at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.getJdbcEnvironmentUsingJdbcMetadata(JdbcEnvironmentInitiator.java:276)
>         at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:107)
>         at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:68)
>         at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:130)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:263)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:238)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:215)
>         at org.hibernate.engine.jdbc.internal.JdbcServicesImpl.configure(JdbcServicesImpl.java:52)
>         at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.configureService(StandardServiceRegistryImpl.java:136)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:247)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:215)
>         at org.hibernate.boot.internal.SessionFactoryOptionsBuilder.<init>(SessionFactoryOptionsBuilder.java:276)
>         at io.quarkus.hibernate.orm.runtime.recording.PrevalidatedQuarkusMetadata.buildSessionFactoryOptionsBuilder(PrevalidatedQuarkusMetadata.java:70)
>         at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:84)
>         at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:72)
>         at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
>         at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
>         at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:157)
>         at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:64)
>         at java.base/java.lang.Thread.run(Thread.java:1583)
> Caused by: org.postgresql.util.PSQLException: The connection attempt failed.
>         at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:358)
>         at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:54)
>         at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:273)
>         at org.postgresql.Driver.makeConnection(Driver.java:446)
>         at org.postgresql.Driver.connect(Driver.java:298)
>         at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:225)
>         at io.agroal.pool.Poolless.createConnection(Poolless.java:366)
>         at io.agroal.pool.Poolless.handlerFromSharedCache(Poolless.java:230)
>         at io.agroal.pool.Poolless.getConnection(Poolless.java:205)
>         at io.agroal.pool.DataSource.getConnection(DataSource.java:86)
>         at io.agroal.api.AgroalDataSource_cE8pf5nze0thbOcWBzTLOVGhcQU_Synthetic_ClientProxy.getConnection(Unknown Source)
>         at io.quarkus.hibernate.orm.runtime.customized.QuarkusConnectionProvider.getConnection(QuarkusConnectionProvider.java:23)
>         at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$ConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:428)
>         at org.hibernate.resource.transaction.backend.jta.internal.JtaIsolationDelegate.doTheWork(JtaIsolationDelegate.java:180)
>         ... 23 more
> Caused by: java.net.SocketTimeoutException: Connect timed out
>         at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:546)
>         at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:592)
>         at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
>         at java.base/java.net.Socket.connect(Socket.java:751)
>         at org.postgresql.core.PGStream.createSocket(PGStream.java:243)
>         at org.postgresql.core.PGStream.<init>(PGStream.java:98)
>         at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:136)
>         at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:262)
>         ... 36 more
> 
> 2024-11-06 21:44:28,825 INFO  [org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider] (main) Initializing database schema. Using changelog META-INF/jpa-changelog-master.xml
> 
> UPDATE SUMMARY
> Run:                        134
> Previously run:               0
> Filtered out:                 0
> -------------------------------
> Total change sets:          134
> 
> 2024-11-06 21:44:46,631 INFO  [org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory] (main) Node name: e02503325114-10438, Site name: null
> 2024-11-06 21:44:46,970 INFO  [org.keycloak.broker.provider.AbstractIdentityProviderMapper] (main) Registering class org.keycloak.broker.provider.mappersync.ConfigSyncEventListener
> 2024-11-06 21:44:47,204 INFO  [org.keycloak.services] (main) KC-SERVICES0050: Initializing master realm
> 2024-11-06 21:44:50,793 INFO  [org.keycloak.exportimport.dir.DirImportProvider] (main) Importing from directory /opt/keycloak/bin/../data/import
> 2024-11-06 21:44:50,793 INFO  [org.keycloak.services] (main) KC-SERVICES0030: Full model import requested. Strategy: IGNORE_EXISTING
> 2024-11-06 21:44:56,025 INFO  [org.keycloak.exportimport.util.ImportUtils] (main) Realm 'brzo' imported
> 2024-11-06 21:44:57,005 INFO  [org.keycloak.services] (main) KC-SERVICES0032: Import finished successfully
> 2024-11-06 21:44:57,560 INFO  [org.keycloak.services] (main) KC-SERVICES0009: Added user 'admin' to realm 'master'
> 2024-11-06 21:44:57,657 INFO  [io.quarkus] (main) Keycloak 25.0.6 on JVM (powered by Quarkus 3.8.5) started in 44.562s. Listening on: http://0.0.0.0:8180. Management interface listening on http://0.0.0.0:9000.
> 2024-11-06 21:44:57,657 INFO  [io.quarkus] (main) Profile prod activated. 
> 2024-11-06 21:44:57,657 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, jdbc-postgresql, keycloak, logging-gelf, micrometer, narayana-jta, reactive-routes, resteasy-reactive, resteasy-reactive-jackson, smallrye-context-propagation, smallrye-health, vertx]
> 2024-11-06 21:45:37,639 INFO  [org.keycloak.events] (executor-thread-1) type="LOGIN", realmId="497e3a7b-a880-42ad-abd7-4c5ce2adf6b3", realmName="master", clientId="security-admin-console", userId="1bec3c63-4b86-4b65-9f2d-86760025ceea", sessionId="8ae8c5f5-8388-4eeb-b36d-daac1d588607", ipAddress="172.18.0.1", auth_method="openid-connect", auth_type="code", response_type="code", redirect_uri="http://localhost:8180/admin/master/console/", consent="no_consent_required", code_id="8ae8c5f5-8388-4eeb-b36d-daac1d588607", username="admin", response_mode="query", authSessionParentId="8ae8c5f5-8388-4eeb-b36d-daac1d588607", authSessionTabId="CWrPRKPqDU8"
> 2024-11-06 21:45:38,621 INFO  [org.keycloak.events] (executor-thread-1) type="CODE_TO_TOKEN", realmId="497e3a7b-a880-42ad-abd7-4c5ce2adf6b3", realmName="master", clientId="security-admin-console", userId="1bec3c63-4b86-4b65-9f2d-86760025ceea", sessionId="8ae8c5f5-8388-4eeb-b36d-daac1d588607", ipAddress="172.18.0.1", token_id="ce408563-3472-402e-9c79-1c53090f4cba", grant_type="authorization_code", refresh_token_type="Refresh", scope="openid profile email", refresh_token_id="0dd79432-3e0f-44e2-acad-a8848f30ae75", code_id="8ae8c5f5-8388-4eeb-b36d-daac1d588607", client_auth_method="client-secret"
> 2024-11-06 21:46:00,105 WARN  [io.agroal.pool] (executor-thread-2) Datasource 'users': The connection attempt failed.
> 2024-11-06 21:46:00,105 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-2) SQL Error: 0, SQLState: 08001
> 2024-11-06 21:46:00,105 **ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-2) The connection attempt failed.**

**Does anybody see what is wrong in the Docker-compose.yml and\or Docker file?
Or is this not possible, and do I need nginx server?**

**Comment by nkamp** [comment]
Some extra information,
Is it possible that it has something todo with network?
I mean if I do "Docker network ls" , the result is:
    
> NETWORK ID     NAME                     DRIVER    SCOPE
> 77de2668e32b   bridge                   bridge    local
> 20a9060dbc6b   host                     host      local
> 7cb3e06f9489   **local_setup_pg-network   bridge    local**   <== local_setup_pg-network instead of pg-network!??
> 11cca9789ed4   none                     null      local

But in my docker compose the network is pg-network!

If I take out the keycloak part from the Docker-compose.yml file and I execute the next commands from the commandline from host laptop, with the same Dockerfile-keycloak:
```
docker build -t keycloak:1 .
docker run --network local_setup_pg-network -p 8180:8180 keycloak:1
```

Than keycloak can make connection with the webapp database, while the network is '`--network local_setup_pg-network`'. Is that not strange? 

And if I execute the second command with '`--network pg-network`' instead of '`--network local_setup_pg-network`'
Than keycloak can't make connection with the postgres webapp database and I get the same error message!

Now I'm asking myself what is missing in the docker compose file?  Or what do I wrong and\or not understand?

**Comment by fostermi** [comment]
Typically `docker-compose` will create a network named after the project name with `_default` appended to it. I'm not sure where `local_setup` is coming from. However, I believe you can override the name by setting the `name` parameter as described here: https://docs.docker.com/reference/compose-file/networks/

**Comment by nkamp** [comment]
I'm still busy to investigate and to get it work. On this moment there is less pressure because I have to fix first other stories. 
 
The 'local_setup' comes from the directory were I have put my docker-compose.yml! So Docker is adding the directory to the network. I don't know why but somebody told me this on another forum. He gave me also some documentation, that I should first read, to understand the Dockerfile vs. Docker-compose. I thought that I was already so far....

Maybe I have to add Nginx server for a reverse-proxy to get the webapplication (Kotlin-Vue) locally work with Keycloak (but I'm not quit sure)
